src/dimension_slice.c

```diff
@@ -983,6 +1025,220 @@ ts_dimension_slice_insert_multi(DimensionSlice **slices, Size num_slices)
+	rel = table_open(catalog_get_table_id(catalog, DIMENSION_SLICE), RowExclusiveLock);
+
+	for (i = 0; i < hs->num_dimensions; i++)
+	{
+		Dimension *dim = &hs->dimensions[i];
+		DimensionSlice *slice;
+		DimensionVec *vec;
+
+		/* Check if an entry already exists first */
+		vec = ts_dimension_slice_equal_scan(dim->fd.id,
+											DIMENSION_SLICE_MINVALUE,
+											DIMENSION_SLICE_MAXVALUE);
+		/* Add a new entry if none exists */
+		if (vec->num_slices == 0)
+		{
+			slice = ts_dimension_slice_create(dim->fd.id,
+											  DIMENSION_SLICE_MINVALUE,
+											  DIMENSION_SLICE_MAXVALUE);
+			dimension_slice_insert_relation(rel, slice);
+		}
+		else
+		{
+			Assert(vec->num_slices == 1);
+			slice = vec->slices[0];
+		}
+
+		ts_chunk_constraints_add(chunk->constraints,
+								 chunk->fd.id,
+								 slice->fd.id,
+								 NULL,
+								 NULL,
+								 DIMENSION_TYPE_CORRELATED);
+	}
+
+	table_close(rel, RowExclusiveLock);
+
+	return i;
+}
+
+/*
+ * Update correlated constraints dimension slices in the catalog for the
+ * provided chunk (it's assumed that the chunk is locked
+ * appropriately).
+ *
+ * Calculate actual ranges for the given chunk for correlated
+ * constraints and update these entries. This allows for the
+ * chunk to be picked up when queries use correlated columns in
+ * WHERE clauses with these ranges.
+ *
+ * If "reset" is specified then we change the range back to MINVALUE
+ * and MAXVALUE
+ *
+ * Returns the number of slices updated.
+ */
+int
+ts_correlated_constraints_dimension_slice_calculate_update(const Hypertable *ht, Chunk *chunk,
+														   bool reset)
+{
+	Size i = 0;
+	Hyperspace *hs = ht->correlated_space;
+	ChunkConstraints *constraints = chunk->constraints;
+	MemoryContext work_mcxt, orig_mcxt;
+
+	/* quick check for correlated constraints. Bail out early if none */
+	if (hs == NULL)
+		return i;
+
+	work_mcxt =
+		AllocSetContextCreate(CurrentMemoryContext, "dimension-slice-work", ALLOCSET_DEFAULT_SIZES);
+	orig_mcxt = MemoryContextSwitchTo(work_mcxt);
+
+	for (int constraint_index = 0; constraint_index < constraints->num_constraints;
+		 constraint_index++)
+	{
+		int slice_id;
+		DimensionSlice *slice;
+		ChunkConstraint *constraint = &constraints->constraints[constraint_index];
+		Datum minmax[2];
+		AttrNumber attno;
+		TupleInfo *ti;
+		ScanIterator slice_iterator;
+		const Dimension *dim;
+
+		if (!is_correlated_constraint(constraint))
+			continue;
+
+		/*
+		 * Find the slice by id. Don't have to lock it because the chunk is
+		 * assumed to be locked when this function is called.
+		 */
+		slice_id = constraint->fd.dimension_slice_id;
+		slice_iterator = ts_dimension_slice_scan_iterator_create(NULL, orig_mcxt);
+		ts_dimension_slice_scan_iterator_set_slice_id(&slice_iterator, slice_id, NULL);
+		ts_scan_iterator_start_scan(&slice_iterator);
+		ti = ts_scan_iterator_next(&slice_iterator);
+		if (ti == NULL)
+			ereport(ERROR, errmsg("dimension slice %d not found!", slice_id));
```

Comment:
AFAICT, this is supposed to exist, so maybe `Ensure` is more appropriate here.
```suggestion
		Ensure(ti != NULL, "dimension slice %d not found!", slice_id);
```

Reply:
done.